### PR TITLE
[EXPERIMENTAL] Traitors are assigned after 3 to 5 minutes

### PR DESCRIPTION
**Pull request by Bibby0110**
# Github documenting your Pull Request

You're no longer safe from becoming a target if you latejoin.
Hopefully this helps increase the amount of people that ready up?

Untested and need to do other changes

- [x] Verify how objectives are picked
- [x] Testing everything

# Changelog

:cl:  
experimental: Traitors are only made aware of their traitor ways after approximately 2 minutes.
/:cl:

### CODE REVIEW

**Comment by adamsong** [comment]
This would prevent people who roll jobs that can't be traitor from rolling traitor

**Comment by adamsong** [comment]
Also, would this negate quiet mode?
EDIT: I see you added a check for it

**Comment by Bibby0110** [comment]
> This would prevent people who roll jobs that can't be traitor from rolling traitor

This works in the exact same way that we do traitors now, the timer is just 2 minutes instead of 1-10 seconds.
My code addition is to re-roll a traitor if they’ve disconnected/died before they’re told they’re a traitor

**Comment by Cartlord** [comment]
But why?
If someone is latejoining so they don't roll tot, I'm pretty sure that means they _don't want to be tot_ - so why force them to be tot anyways? What's the gain?

File: code/game/gamemodes/traitor/traitor.dm [hunk 1]
```diff
@@ -77,14 +77,15 @@
 /datum/game_mode/traitor/post_setup()
 	for(var/datum/mind/traitor in pre_traitors)
 		var/datum/antagonist/traitor/new_antag = new antag_datum()
```
Comment (adamsong) [inline code/game/gamemodes/traitor/traitor.dm]: This variable is now unused

File: code/game/gamemodes/traitor/traitor.dm [hunk 2]
```diff
@@ -99,6 +100,46 @@
...
+			continue
+		if(!applicant.mind)
+			continue
+		if(!applicant.stat != CONSCIOUS)
+			continue
+		if(applicant.mind.assigned_role in protected_jobs) 
+			continue
+		if(applicant.mind.assigned_role in restricted_jobs) 
+			continue
+		if(applicant.mind.quiet_mode)
```
Comment (adamsong) [inline code/game/gamemodes/traitor/traitor.dm]: `quiet_round` is the variable you are looking for

File: code/game/gamemodes/traitor/traitor.dm [hunk 3]
```diff
@@ -99,6 +100,46 @@
...
+			continue
+		if(HAS_TRAIT(applicant, TRAIT_MINDSHIELD))
+			continue
+		if(is_banned_from(applicant.ckey, list(antag_flag, ROLE_SYNDICATE)))
+			continue
+		if(!(role in applicant.client.prefs.be_special))
```
Comment (adamsong) [inline code/game/gamemodes/traitor/traitor.dm]: role should probably be `ROLE_TRAITOR`

**Review by adamsong — CHANGES_REQUESTED** [review]
Looks mostly good, just some small changes. Get this compiling and it can be test merged. I will definitely be interested to see how well it works.

As for the objectives, they are rolled when the datum is given, so in this case after the two minutes.

File: code/game/gamemodes/traitor/traitor.dm (hunk 2 above)
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
		if(applicant.mind.quiet_round)
```

File: code/game/gamemodes/traitor/traitor.dm (hunk 1 above)
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
```

File: code/game/gamemodes/traitor/traitor.dm (hunk 3 above)
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
		if(!(antag_flag in applicant.client.prefs.be_special))
```

**Comment by adamsong** [comment]
Does this apply to all traitor gamemodes, such as traitorlings, or just solo traitors?

**Comment by Bibby0110** [comment]
> Does this apply to all traitor gamemodes, such as traitorlings, or just solo traitors?

just sole traitor for now

**Review by adamsong — DISMISSED** [review]

File: code/game/gamemodes/traitor/traitor.dm [hunk 4]
```diff
@@ -76,15 +76,15 @@
 
 /datum/game_mode/traitor/post_setup()
 	for(var/datum/mind/traitor in pre_traitors)
-		var/datum/antagonist/traitor/new_antag = new antag_datum()
-		addtimer(CALLBACK(traitor, /datum/mind.proc/add_antag_datum, new_antag), rand(10,100))
+		addtimer(CALLBACK(src, /datum/game_mode/traitor.proc/add_traitor_delayed, traitor), rand(2 MINUTES, (2 MINUTES + 10 SECONDS)))
```
Comment (adamsong) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
		addtimer(CALLBACK(src, /datum/game_mode/traitor.proc/add_traitor_delayed, traitor), rand(3 MINUTES, 5 MINUTES))
```

**Review by adamsong — COMMENTED** [review]
Add a greater variance to the start delay

File: code/game/gamemodes/traitor/traitor.dm [hunk 5]
```diff
@@ -76,15 +76,15 @@
 
 /datum/game_mode/traitor/post_setup()
 	for(var/datum/mind/traitor in pre_traitors)
-		var/datum/antagonist/traitor/new_antag = new antag_datum()
-		addtimer(CALLBACK(traitor, /datum/mind.proc/add_antag_datum, new_antag), rand(10,100))
+		addtimer(CALLBACK(src, /datum/game_mode/traitor.proc/add_traitor_delayed, traitor), rand(3 MINUTES, 5 MINUTES))
```
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
		addtimer(CALLBACK(src, /datum/game_mode/traitor.proc/add_traitor_delayed, traitor), rand(2 MINUTES, (2 MINUTES + 10 SECONDS)))
```

File: code/game/gamemodes/traitor/traitor.dm [hunk 6]
```diff
@@ -76,15 +76,15 @@
...
+		
 	if(!exchange_blue)
 		exchange_blue = -1 //Block latejoiners from getting exchange objectives
 	..()
 
 	//We're not actually ready until all traitors are assigned.
 	gamemode_ready = FALSE
-	addtimer(VARSET_CALLBACK(src, gamemode_ready, TRUE), 101)
+	addtimer(VARSET_CALLBACK(src, gamemode_ready, TRUE), (2 MINUTES + 11 SECONDS))
```
Comment (adamsong) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
	addtimer(VARSET_CALLBACK(src, gamemode_ready, TRUE), (5 MINUTES + 1 SECONDS))
```

File: code/game/gamemodes/traitor/traitor.dm (hunk 4 above)
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
		addtimer(CALLBACK(src, /datum/game_mode/traitor.proc/add_traitor_delayed, traitor), rand(3 MINUTES, (5 MINUTES + 10 SECONDS)))
```

File: code/game/gamemodes/traitor/traitor.dm [hunk 7]
```diff
@@ -76,15 +76,15 @@
...
+		addtimer(CALLBACK(src, /datum/game_mode/traitor.proc/add_traitor_delayed, traitor), rand(2 MINUTES, (2 MINUTES + 10 SECONDS)))
+		
 	if(!exchange_blue)
 		exchange_blue = -1 //Block latejoiners from getting exchange objectives
 	..()
 
 	//We're not actually ready until all traitors are assigned.
 	gamemode_ready = FALSE
-	addtimer(VARSET_CALLBACK(src, gamemode_ready, TRUE), 101)
+	addtimer(VARSET_CALLBACK(src, gamemode_ready, TRUE), (2 MINUTES + 11 SECONDS))
```
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
	addtimer(VARSET_CALLBACK(src, gamemode_ready, TRUE), (5 MINUTES + 11 SECONDS))
```

File: code/game/gamemodes/traitor/traitor.dm [hunk 8]
```diff
@@ -99,6 +99,46 @@
 					if(!(character.job in restricted_jobs))
 						add_latejoin_traitor(character.mind)
 
+/datum/game_mode/traitor/proc/add_traitor_delayed(datum/mind/traitor)
+	if(!traitor.current.client || (traitor.current.stat != CONSCIOUS))
```
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm]:
```suggestion
	if(!traitor || !traitor.current || !traitor.current.client || (traitor.current.stat != CONSCIOUS) || istype(traitor.current.loc, /obj/machinery/cryopod))
```

File: code/game/gamemodes/traitor/traitor.dm [hunk 9]
```diff
@@ -99,6 +99,46 @@
...
+			continue
+		if(is_banned_from(applicant.ckey, list(antag_flag, ROLE_SYNDICATE)))
+			continue
+		if(!(antag_flag in applicant.client.prefs.be_special))
+			continue
+		if(!age_check(applicant.client))
+			continue
+		potential_candidates += applicant
+	if(!potential_candidates)
+		return
```
Comment (Bibby0110) [inline code/game/gamemodes/traitor/traitor.dm:137]:
```suggestion
		message_admins("Failed to find new antag after original one left! Check the antag balance please.")
		return
```

**Comment by monster860** [comment]
Time to just wait 5 minutes after latejoining instead

**Comment by Bibby0110** [comment]
> Time to just wait 5 minutes after latejoining instead

most people probably won't do this so we're good

**Comment by adamsong** [comment]
Seems to be working in test merge, need to do a bit more testing, and then add this to the combined modes

**Comment by MegaEmpirical** [comment]
Nifty

**Comment by Cartlord** [comment]
BRO THIS ROLLED A FREE MINER AS SYNDIE ROUND 36291

**Comment by adamsong** [comment]
Seems to also work on traitorling rounds, possibly other combo rounds

**Comment by Cartlord** [comment]
So uh, there doesn't seem to be anything in place that keeps Lifegivers, Free Miners, or Charlie Station folks from being antags. That seems like a bad thing - what's a Lifegiver meant to do if they get Hijack or something?

**Comment by adamsong** [comment]
Add a check to make sure they are on the station z-level, so if a ghost role is on the station, such as free miners, they can get it, and it means they can get on the station for their objectives, otherwise they can't become traitor

File: code/game/gamemodes/traitor/traitor.dm [hunk 10]
```diff
@@ -99,6 +99,47 @@
...
+	if(!traitor || !traitor.current || !traitor.current.client || (traitor.current.stat != CONSCIOUS) || istype(traitor.current.loc, /obj/machinery/cryopod))
+		create_new_traitor()
+		return
+	var/datum/antagonist/traitor/new_antag = new antag_datum()
+	traitor.add_antag_datum(new_antag)
+
+/datum/game_mode/traitor/proc/create_new_traitor()
+	var/list/potential_candidates = list()
+	for(var/mob/living/carbon/human/applicant in GLOB.player_list)
+		if(!applicant.client)
```
Comment (adamsong) [inline code/game/gamemodes/traitor/traitor.dm:112]:
```suggestion
		if(!is_station_level(applicant.z))
			continue
		if(!applicant.client)
```

**Review by adamsong — DISMISSED** [review]

**Review by JamieD1 — APPROVED** [review]

**Pull request merged**